server/src/main/java/org/elasticsearch/common/util/LazyCopyOnWriteMap.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.routing.allocation.allocator.DesiredBalanceComputer;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This map is designed to be constructed from an immutable map and only be copied if a (rare) mutation operation occurs.
+ * It could be converted back to an immutable map using `org.elasticsearch.common.util.LazyCopyOnWriteMap#toImmutableMap()`.
+ */
+public class LazyCopyOnWriteMap<K, V> implements Map<K, V> {
```

Comment:
Naming nit/suggestion: no need to call this `Lazy` but maybe `CopyOnWriteMap` implies too much thread-safety.

```suggestion
public class CopyOnFirstWriteMap<K, V> implements Map<K, V> {
```